## Pre-lecture 

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [2]:
df

row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   aye aye   villager-admiral   
1            DJ K.K.  sidekick   villager-agent-s   
2         K.K. House   snuffle     villager-agnes   
3         Steep Hill   Ayyeeee        villager-al   
4        Forest Life  it'sa me   villager-alfonso   
..               ...       ...                ...   
386         My Place    hay-OK    villager-winnie   
387        K.K. Song   snarrrl  villager-wolfgang   
388     Soulful K.K.   tsk tsk      villager-yuka   
389         K.K. D&B     pronk      villager-zell   
390  Spring Blossoms     bloop    villager-zucker   

                                                   url  
0    https://villagerdb.com/images/villagers/thumb/...  
1    https://villagerdb.com/images/villagers/thumb/...  
2    https://villagerdb.com/images/villagers/thumb/...  
3    https://villagerdb.com/images/villagers/thumb/...  
4    https://villagerdb.com/images/villagers/thumb/...  
..                                                 ...  
386  https://villagerdb.com/images/villagers/thumb/...  
387  https://villagerdb.com/images/villagers/thumb/...  
388  https://villagerdb.com/images/villagers/thumb/...  
389  https://villagerdb.com/images/villagers/thumb/...  
390  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 11 columns]

In [3]:
num_rows = df.shape[0]
num_columns = df.shape[1]
print('Number of rows: ', num_rows)
print('Number of columns: ', num_columns)

Number of rows:  391
Number of columns:  11


In [4]:
df.columns

Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')

## Observations and Variables

An observation is an fact or figure we collect about a given variable. In the context of the dataset, the observation refers to each row which contains the information of the villager (such as its name, species, and personality).

A variable is any characteristic, number, or quantity that can be measured or counted to describe each observation. There are numerical variables (values with equal units and scales) and categorical variables (variable identifies category or group, not order or scale). In the context of the dataset, the variable refers to the columns. Names, species, birthdays, personality types, and other traits of the villagers are listed in the columns. These columns give you data points for each villager.

## To summarize the dataset

In [5]:
# General
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB


In [6]:
# Summarizing Categorical Variables
df.describe(include=['object'])

id     name gender species birthday personality          song  \
count       390      391    391     391      391         391           380   
unique      390      391      2      35      361           8            92   
top     admiral  Admiral   male     cat     1-27        lazy  K.K. Country   
freq          1        1    204      23        2          60            10   

         phrase           full_id  \
count       391               391   
unique      388               391   
top     wee one  villager-admiral   
freq          2                 1   

                                                      url  
count                                                 391  
unique                                                391  
top     https://villagerdb.com/images/villagers/thumb/...  
freq                                                    1

In [7]:
# Summarizing Numerical Variables
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [8]:
# The number of each species
species_counts = df['species'].value_counts()
print(species_counts)

# The number of each personality type
personality_counts = df['personality'].value_counts()
print(personality_counts)


species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64
personality
lazy      60
normal    59
cranky    55
snooty    55
jock      55
peppy     49
smug      34
uchi      24
Name: count, dtype: int64


## Key Discrepancies:

In [9]:
url_1 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
sb = pd.read_csv(url_1)
sb.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [10]:
sb

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [11]:
num_rows_1 = sb.shape[0]
num_columns_1 = sb.shape[1]
print('Number of rows: ', num_rows_1)
print('Number of columns: ', num_columns_1)

Number of rows:  891
Number of columns:  15


In [12]:
sb.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [13]:
# General
sb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB


In [14]:
# Summarizing Categorical Variables
sb.describe(include=['object'])

sex embarked  class  who deck  embark_town alive
count    891      889    891  891  203          889   891
unique     2        3      3    3    7            3     2
top     male        S  Third  man    C  Southampton    no
freq     577      644    491  537   59          644   549

In [15]:
# Summarizing Numerical Variables
sb.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

sb.shape gives us the total number of rows and columns of the dataset, while the "count" value from sb.describe() for each column represents the number of non-null entries in that column. 

In the above chart, the count values for "survived", "pclass", "sibsp", "parch", and "fare" are 891, which matches the total number of rows returned by sb.shape. The count value for "age" is 714, which differs from 891 because the "age" column has 177 missing values.

## The difference between an "attribute" and a "method"

Summary: In Python, attributes are variables that are a part of an object and hold data on its traits and properties. They can be used to depict information about the thing, such as facts or details. They can be accessed without (), and no action or computation is performed. Methods are functions that are connected to an object or class. They are employed to carry out particular actions or activities on the data kept in objects or classes. They can be called with (), and they can perform operations on the object.


## Summaries of ChatBot Sessions

Here’s a summary of our interactions related to your homework assignment on data analysis using Python and pandas:

---

### 1. **Analyzing the Titanic Dataset**

- **Dataset URL**: https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv
- We discussed how to load and explore this dataset using pandas.
  
  **Code to load the dataset**:
  ```python
  import pandas as pd
  url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
  df = pd.read_csv(url)
  ```

### 2. **Dimensions of the Dataset**

- To determine the dimensions (number of rows and columns) of the dataset, we used:
  ```python
  df.shape
  ```
  This returns the tuple `(rows, columns)`, where rows represent the observations, and columns represent the variables.

### 3. **Discrepancy Between `df.shape` and `df.describe()`**

- **`df.shape`** gives the total number of rows and columns in the dataset.
- **`df.describe()`** by default summarizes **only numeric columns** and reports statistical values like count, mean, standard deviation, min, max, etc.
  
  - If a column has missing values, the count reported in `df.describe()` will be lower than the total number of rows in `df.shape`.
  - **Non-numeric columns** (like categorical data) are excluded from `df.describe()` unless explicitly included using:
    ```python
    df.describe(include='all')
    ```

  **Example Summary Code**:
  ```python
  df.describe(include='all')
  ```

  This command will include both numeric and categorical variables in the summary.

### 4. **Categorical vs Numeric Variables**

- We discussed the difference between **observations** (rows) and **variables** (columns), where variables could be **numeric** (e.g., `age`, `fare`) or **categorical** (e.g., `sex`, `embarked`).

---

This summary should cover the key aspects of your dataset exploration and analysis using pandas. Let me know if any details need further clarification!

## Link

https://chatgpt.com/share/9d4bcec3-75b1-4f48-a6a3-224107da1055

## Post-lecture

## Definitions

Count: The number of non-missing (non-null) values in the column.

Mean: The average of the values. It is the sum of all values divided by the number of non-missing values.

Std: Standard deviation. A measure of how spread out the values are. It represents the average distance between each value and the mean.

Min: The smallest value in the column.

Max: The largest value in the column.

Median(Md): score corresponding to the point having 50% of the observations below it. It is also the second quartile (50%).

Queartiles divide an ordered (lowest to highest) dataset into quarters. 25%, 50%, and 75% are the first, second, and third quantiles referred to as  𝑄1, the median, and  𝑄3

## df.dropna() or del df['col']

Prefer df.dropna()

If the dataset contains missing values sporadically throughout multiple columns but only affects a small number of rows, we may prefer df.dropna() over df['col'] because it allows us to keep as many columns as possible and only remove rows that have missing data.

Example: 

In [16]:
print(max(sb.isna().sum(axis=1)))
sb

2


survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In the sb dataset, for rows that have 2 missing values, we may choose to drop the rows rather than delete the columns.

Prefer del['col']

del['col'] may be chosen when a whole column has a large quantity of missing data. It would be preferable to eliminate the column in this situation in order to prevent adding bias or mistakes into the analysis.

Example:

In [17]:
sb.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In the sb dataset, column 'deck' contains 688 missing values. We may choose to delete this column rather than drop rows.

Importance of Applying del df['col'] Before df.dropna(): Prior to using df.dropna(), it could be more effective to remove any columns that have a high number of missing data. 

## Removing all the missing values

In [18]:
before_report = sb.isnull().sum()

Step 1: Delete the Age column and Deck column (which contains many missing values):

In [19]:
del sb['age']
del sb['deck']
sb

survived  pclass     sex  sibsp  parch     fare embarked   class    who  \
0           0       3    male      1      0   7.2500        S   Third    man   
1           1       1  female      1      0  71.2833        C   First  woman   
2           1       3  female      0      0   7.9250        S   Third  woman   
3           1       1  female      1      0  53.1000        S   First  woman   
4           0       3    male      0      0   8.0500        S   Third    man   
..        ...     ...     ...    ...    ...      ...      ...     ...    ...   
886         0       2    male      0      0  13.0000        S  Second    man   
887         1       1  female      0      0  30.0000        S   First  woman   
888         0       3  female      1      2  23.4500        S   Third  woman   
889         1       1    male      0      0  30.0000        C   First    man   
890         0       3    male      0      0   7.7500        Q   Third    man   

     adult_male  embark_town alive  alone  
0          True  Southampton    no  False  
1         False    Cherbourg   yes  False  
2         False  Southampton   yes   True  
3         False  Southampton   yes  False  
4          True  Southampton    no   True  
..          ...          ...   ...    ...  
886        True  Southampton    no   True  
887       False  Southampton   yes   True  
888       False  Southampton    no  False  
889        True    Cherbourg   yes   True  
890        True   Queenstown    no   True  

[891 rows x 13 columns]

Step 2: Drop rows with missing data in the remaining columns:

In [20]:
sb_cleaned = sb.dropna()
sb_cleaned

survived  pclass     sex  sibsp  parch     fare embarked   class    who  \
0           0       3    male      1      0   7.2500        S   Third    man   
1           1       1  female      1      0  71.2833        C   First  woman   
2           1       3  female      0      0   7.9250        S   Third  woman   
3           1       1  female      1      0  53.1000        S   First  woman   
4           0       3    male      0      0   8.0500        S   Third    man   
..        ...     ...     ...    ...    ...      ...      ...     ...    ...   
886         0       2    male      0      0  13.0000        S  Second    man   
887         1       1  female      0      0  30.0000        S   First  woman   
888         0       3  female      1      2  23.4500        S   Third  woman   
889         1       1    male      0      0  30.0000        C   First    man   
890         0       3    male      0      0   7.7500        Q   Third    man   

     adult_male  embark_town alive  alone  
0          True  Southampton    no  False  
1         False    Cherbourg   yes  False  
2         False  Southampton   yes   True  
3         False  Southampton   yes  False  
4          True  Southampton    no   True  
..          ...          ...   ...    ...  
886        True  Southampton    no   True  
887       False  Southampton   yes   True  
888       False  Southampton    no  False  
889        True    Cherbourg   yes   True  
890        True   Queenstown    no   True  

[889 rows x 13 columns]

In [21]:
after_report = sb_cleaned.isnull().sum()


In [22]:
print("Before report of missing values:\n", before_report)
print("After report of missing values:\n", after_report)

Before report of missing values:
 survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64
After report of missing values:
 survived       0
pclass         0
sex            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64


## Troubleshooting

In [23]:
url_2 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
sb = pd.read_csv(url_2)
sb

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [24]:
sb.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [25]:
sb.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [26]:
sb.groupby("pclass")["fare"].describe()


count       mean        std  min       25%      50%   75%       max
pclass                                                                     
1       216.0  84.154687  78.380373  0.0  30.92395  60.2875  93.5  512.3292
2       184.0  20.662183  13.417399  0.0  13.00000  14.2500  26.0   73.5000
3       491.0  13.675550  11.778142  0.0   7.75000   8.0500  15.5   69.5500

In [27]:
sb.groupby("sex")["age"].describe()


count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

The count from df.describe() provides a global picture of the missingness. It aids in understanding the quantity of data contained for each column in the dataset.

The count in df.groupby("col1")["col2"].describe() serves as a more precise indicator of the quantity of data accessible inside each category.

 Question 9: Yes

## Summary

Sure! Here’s a summary of our exchanges:

---

**Summary of Exchanges:**

1. **Initial Code Issue:**
   - **Error:** `NameError: name 'pd' is not defined`
   - **Cause:** The `pandas` library was not imported.
   - **Solution:** Import `pandas` using `import pandas as pd`.

2. **File Download Error:**
   - **Error:** `HTTPError: HTTP Error 404: Not Found`
   - **Cause:** Incorrect URL used (`titantics.csv` instead of `titanic.csv`).
   - **Solution:** Correct the URL to `https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv`.

3. **Syntax Error:**
   - **Error:** `SyntaxError: '(' was never closed`
   - **Cause:** Missing closing parenthesis.
   - **Solution:** Ensure all parentheses are properly closed in `pd.read_csv(url_2)`.

4. **AttributeError:**
   - **Error:** `AttributeError: 'DataFrame' object has no attribute 'group_by'`
   - **Cause:** Incorrect method name (`group_by` instead of `groupby`).
   - **Solution:** Use `groupby` for grouping the DataFrame.

5. **KeyError:**
   - **Error:** `KeyError: 'Sex'`
   - **Cause:** Column name `'Sex'` does not match (case-sensitive).
   - **Solution:** Use the correct column name, typically `'sex'`, by checking the column names with `sb.columns`.

6. **NameError in Column Selection:**
   - **Error:** `NameError: name 'age' is not defined`
   - **Cause:** Column name `'age'` not in quotes.
   - **Solution:** Use `"age"` in quotes for selecting columns.

7. **Final Column Selection Error:**
   - **Error:** `NameError: name 'age' is not defined`
   - **Cause:** Column name `'age'` should be in quotes.
   - **Solution:** Correct code: `sb.groupby("sex")["age"].describe()`.

---

This summary captures the issues and solutions discussed during our interactions. Let me know if there's anything else you need!

## Link

https://chatgpt.com/share/89369fbb-7f01-4245-abd7-5f28ab933fdf